### PR TITLE
Loki Query Frontend

### PR DESCRIPTION
**Pull request by cyriltovena**
After many refactoring PRs in Cortex, this PR introduces a `query-frontend` component in loki.
The `query-frontend`  sits in front of queriers and communicates with them to dispatch work, allowing parallelization and queries speed improvement.

I have been running the frontend for a month in dev and it really shows promising result. However it  definitely requires bigger queriers pool compared to Cortex.

The current implementation targets specifically slow queries.

## Metrics queries

Metrics queries like cortex will be aligned, splitted by 4 hours by default and result cached. In fact we are using the same set of middlewares.For some streams 4 hours block can still be too much (high throughput), but I expect the sharding parallelization to split again those 4 hours block into 16 (Future works), in the meantime you can tweak the interval to a lower range.

## Log Regex queries

Log regex queries are splitted also by interval of 4 hours, however compared to cortex they are run in sequence and each block is splitted by 16 (configurable) and run in parallel. This allows to check for every 4 hours if we already have enough result before moving to the next block.

Anything else (non regex log queries, labels, etc...) run the same way but still goes through the frontend. Except for labels names queries I actually don't plan any work since those queries are already very fast. (500~700ms)

The loki jsonnet library has been updated.

PS: I had to update to latest cortex and there were some changes related to the ring and some vendoring issue.

Next steps :
- Add some documentation.
- Implement log result cache. (tricky but should really improve resources consumption)
- Add parallelization via sharding /cc @owen-d 

/cc @slim-bean @tomwilkie @joe-elliott @rfratto 

Here is a screenshot of a trace for a regex queries on an high throughput stream:
image

This is not yet part of the single binary, I still need to think about how we want to design this, it could be very nice to swap grpc for direct querier library calls.

### CODE REVIEW

File: pkg/loghttp/query.go [hunk 1]
```diff
@@ -70,6 +71,18 @@ func (Matrix) Type() ResultType { return ResultTypeMatrix }
 // Streams is a slice of Stream
 type Streams []Stream
 
+func (s Streams) ToProto() []logproto.Stream {
+	if len(s) == 0 {
+		return nil
+	}
+	result := make([]logproto.Stream, 0, len(s))
+	for _, s := range s {
+		entries := *(*[]logproto.Entry)(unsafe.Pointer(&s.Entries))
```
Comment (owen-d) [inline pkg/loghttp/query.go]: Add a test for this to insulate us from struct changes over time

File: production/ksonnet/loki/config.libsonnet [hunk 2]
```diff
@@ -91,13 +93,50 @@
       server: {
         graceful_shutdown_timeout: '5s',
         http_server_idle_timeout: '120s',
-        grpc_server_max_recv_msg_size: 1024 * 1024 * 64,
+        grpc_server_max_recv_msg_size: 100 << 20,
```
Comment (owen-d) [inline production/ksonnet/loki/config.libsonnet]: all these `100 << 20` should probably be kept as a variable somewhere and reused as it seems coordinated in a number of places

File: pkg/querier/queryrange/split_by_interval.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+			resps = append(resps, reqResp.Response)
+		}
+
+		resp, err := s.merger.MergeResponse(resps...)
+		if err != nil {
+			sp.Finish()
+			return nil, err
+		}
+
+		lokiRes := resp.(*LokiResponse)
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/querier/queryrange/split_by_interval.go]: This is potentially a good optimization, but let's keep this for later.

File: pkg/loki/modules.go [hunk 4]
```diff
@@ -261,6 +275,42 @@ func (t *Loki) stopStore() error {
...
+			t.frontend.Handler(),
+		),
+	)
+	return
+}
+
+func (t *Loki) stopQueryFrontend() (err error) {
+	t.frontend.Close()
+	if t.stopper != nil {
+		if err := t.stopper.Stop(); err != nil {
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/loki/modules.go]: Remove the confusion.

File: pkg/loki/loki.go [hunk 5]
```diff
@@ -96,7 +108,19 @@ func (t *Loki) setupAuthMiddleware() {
 			middleware.ServerUserHeaderInterceptor,
 		}
 		t.cfg.Server.GRPCStreamMiddleware = []grpc.StreamServerInterceptor{
-			middleware.StreamServerUserHeaderInterceptor,
+			func(srv interface{}, ss grpc.ServerStream, info *grpc.StreamServerInfo, handler grpc.StreamHandler) error {
+				switch info.FullMethod {
+				// Don't check auth header on TransferChunks, as we weren't originally
```
[the remark this answers is not in the document]
Comment (cyriltovena) [reply pkg/loki/loki.go]: @rfratto Should I add `"/logproto.Ingester/TransferChunks"`, I think it was working before using the `middleware.StreamServerUserHeaderInterceptor` ?

May be just the comment is confusing, I copied it over from cortex.

[the remark this answers is not in the document]
Comment (rfratto) [reply pkg/loki/loki.go]: I don't understand the mechanics but if we want to copy what Cortex is doing, it [does have a case](https://github.com/cortexproject/cortex/blob/9bc49ceb68746ddfda57d5ceaa62caee1bd2e2c8/pkg/cortex/cortex.go#L201) for TransferChunks.

**Comment by cyriltovena** [comment]
dam windows

**Pull request merged**